nexus/mgs-updates/src/driver_update/test_host_phase_1.rs

```diff
@@ -0,0 +1,658 @@
+        override_baseboard_id: None,
+        override_expected_sp_component: ExpectedSpComponent::HostPhase1 {
+            host_phase_2_state: host_phase_2_state.clone(),
+            override_expected_phase_1_slot: None,
+            override_expected_boot_disk: None,
+            override_expected_active_phase_1: None,
+            override_expected_active_phase_2: None,
+            override_expected_inactive_phase_1: None,
+            override_expected_inactive_phase_2: None,
+        },
+        override_progress_timeout: None,
+    };
+
+    let desc2 = UpdateDescription {
+        gwtestctx: &gwtestctx,
+        artifacts: &artifacts,
+        sp_type: SpType::Sled,
+        slot_id: 1,
+        artifact_hash: &artifacts.host_phase_1_artifact_hash,
+        override_baseboard_id: None,
+        override_expected_sp_component: ExpectedSpComponent::HostPhase1 {
+            host_phase_2_state: host_phase_2_state.clone(),
+            override_expected_phase_1_slot: None,
+            override_expected_boot_disk: None,
+            override_expected_active_phase_1: None,
+            override_expected_active_phase_2: None,
+            override_expected_inactive_phase_1: Some(
+                artifacts.host_phase_1_artifact_hash,
+            ),
+            override_expected_inactive_phase_2: None,
+        },
+        override_progress_timeout: None,
+    };
+
+    let mut in_progress1 = desc1.setup().await;
+    let in_progress2 = desc2.setup().await;
+
+    // Start one, but pause it while waiting for the update to upload.
+    in_progress1.run_until_status(UpdateAttemptStatus::UpdateWaiting).await;
+
+    // Run the other; it should complete successfully.
+    let finished2 = in_progress2.finish().await;
+    finished2.expect_host_phase_1_success(UpdateCompletedHow::CompletedUpdate);
+
+    // Now finish the first update.
+    let finished1 = in_progress1.finish().await;
+
+    // Because we're using "reset the SP" as our final host phase 1 update step,
+    // there are two likely ways it can complete in production. It's in its
+    // "poll for upload complete" loop. It might see that update2's upload
+    // completes (if it catches it between the upload completing and the other
+    // update resetting the SP) or it might see a "lost update" error (if its
+    // next-to-last poll sees the upload still in progress and its last poll
+    // sees no upload in progress at all, because of the SP reset).
+    //
+    // The corresponding SP update step pauses the update futures in a way that
+    // it sees the first variant, which results in a success with
+    // `UpdateCompletedHow::WaitedForConcurrentUpdate`. The way we've structured
+    // this test causes us to see the second: this update should fail with
+    // `ApplyUpdateError::SpUpdateLost`.
+    finished1.expect_failure(&|error, _sp1, _sp2| {
+        assert_matches!(error, ApplyUpdateError::SpUpdateLost);
+    });
+
```

Comment:
What do you think of re-running update1 at this point and verifying that it makes no changes?

Reply:
Great idea - added in 33729cee7 (and it does indeed make no changes)